pkg/service/model_builder.go

```diff
@@ -80,17 +92,67 @@ type defaultModelBuilder struct {
 	enableBackendSG          bool
 	disableRestrictedSGRules bool
 
+	stackGlobalCache map[core.StackID]core.Stack
+
 	clusterName         string
 	vpcID               string
 	defaultTags         map[string]string
 	externalManagedTags sets.String
 	defaultSSLPolicy    string
 	defaultTargetType   elbv2model.TargetType
 	enableIPTargetType  bool
+
+	client client.Client
+
+	initialized bool
+	lock        sync.Mutex
 }
 
 func (b *defaultModelBuilder) Build(ctx context.Context, service *corev1.Service) (core.Stack, *elbv2model.LoadBalancer, bool, error) {
-	stack := core.NewDefaultStack(core.StackID(k8s.NamespacedName(service)))
+	// Initialize the global cache if not initialized
+	if !b.initialized {
+		// if not initialized, we need to build the global cache based on existing services
+		var serviceList corev1.ServiceList
+		if err := b.client.List(context.Background(), &serviceList); err != nil {
+			return nil, nil, false, err
+		}
+		for _, svc := range serviceList.Items {
+			if svc.Annotations[LoadBalancerStackKey] != "" && !b.serviceUtils.IsServicePendingFinalization(&svc) {
+				stackID := core.StackID(types.NamespacedName{
+					Namespace: "stack",
+					Name:      svc.Annotations[LoadBalancerStackKey],
+				})
+				if b.stackGlobalCache[stackID] == nil {
+					b.stackGlobalCache[stackID] = core.NewDefaultStack(stackID)
+				}
+				b.stackGlobalCache[stackID].AddService(&svc)
+			}
+		}
+		b.initialized = true
+	}
+
+	// For each stack ID, if we found the stack annotation, this means the service will be sharing the same stack with other services
+	// If so, we should reuse the same stack in the cache so that we can reuse the load balancer with shared listeners
+	stackID := core.StackID(k8s.NamespacedName(service))
+	var stack core.Stack
+	stack = core.NewDefaultStack(stackID)
+	if service.Annotations[LoadBalancerAllocatingPortKey] == "true" {
+		// service will be allocated to a stack. The external controller will be
+		if service.Annotations[LoadBalancerStackKey] == "" {
+			return nil, nil, false, errors.Errorf("service %v/%v is waiting to allocated for a stack", service.Namespace, service.Name)
+		}
+		stackID = core.StackID(types.NamespacedName{
+			Namespace: "stack",
+			Name:      service.Annotations[LoadBalancerStackKey],
+		})
+		if b.stackGlobalCache[stackID] == nil {
```

Comment:
Reads also need to be protected by a (read) lock so that concurrent reads and writes don't happen.

Reply:
addressed 👍

Reply:
You protected the read below, but not this read.